File: lib/traffic-light-node.ts
"use client"

import type { TrafficLightServer } from "./traffic-light-server"

export enum LightState {
  RED = "RED",
  GREEN = "GREEN",
  YELLOW = "YELLOW",
}

export class TrafficLightNode {
  private readonly HEARTBEAT_INTERVAL = 3000 // 3 seconds
  private readonly BASE_RED_DURATION = 10000 // Base duration
  private readonly BASE_GREEN_DURATION = 10000 // Base duration
  private readonly BASE_YELLOW_DURATION = 3000 // Fixed yellow duration

  private active = true
  private server: TrafficLightServer
  private heartbeatInterval: NodeJS.Timeout | null = null
  private stateUpdateInterval: NodeJS.Timeout | null = null
  private logCallback: (log: string) => void
  private neighbors: Set<string> = new Set() // Store neighbor node IDs
  private failedNeighbors: Set<string> = new Set() // Track failed neighbors

  public nodeId: string
  public currentState: LightState
  public isDisabled: boolean = false
  public lastStateChange: number = Date.now()
  public onHeartbeat?: (nodeId: string) => void

  // Dynamic durations based on network state
  public RED_DURATION: number = 10000
  public GREEN_DURATION: number = 10000
  public readonly YELLOW_DURATION: number = 3000 // Yellow remains fixed for safety

  constructor(nodeId: string, server: TrafficLightServer, logCallback: (log: string) => void) {
    this.nodeId = nodeId
    this.server = server
    this.logCallback = logCallback

    // Set initial state based on node ID to create a pattern
    if (nodeId === "Node1") {
      this.currentState = LightState.RED
    } else if (nodeId === "Node2") {
      this.currentState = LightState.GREEN
    } else if (nodeId === "Node3") {
      this.currentState = LightState.RED
    } else {
      // For additional nodes, set initial state based on position
      const nodeNum = Number.parseInt(nodeId.substring(4))
      this.currentState = nodeNum % 3 === 2 ? LightState.GREEN : LightState.RED
    }

    this.log(`Traffic light initialized to ${this.currentState}`)
    this.start()
  }

  private start(): void {
    this.startHeartbeatSender()
    this.startTrafficLightStateUpdater()
  }

  private startHeartbeatSender(): void {
    this.heartbeatInterval = setInterval(() => {
      if (this.active) {
        this.sendHeartbeat()
      }
    }, this.HEARTBEAT_INTERVAL)

    // Send initial heartbeat
    this.sendHeartbeat()
  }

  private sendHeartbeat(): void {
    this.server.receiveHeartbeat(this.nodeId)
    this.log(`Sent heartbeat from node ${this.nodeId}`)
    this.onHeartbeat?.(this.nodeId)
  }

  private startTrafficLightStateUpdater(): void {
    const updateState = async () => {
      if (!this.active) return

      switch (this.currentState) {
        case LightState.RED:
          await this.delay(this.RED_DURATION)
          if (this.active) {
            this.currentState = LightState.GREEN
            this.lastStateChange = Date.now()
            this.log(`Traffic light changed to ${this.currentState}`)
          }
          break

        case LightState.GREEN:
          await this.delay(this.GREEN_DURATION)
          if (this.active) {
            this.currentState = LightState.YELLOW
            this.lastStateChange = Date.now()
            this.log(`Traffic light changed to ${this.currentState}`)
          }
          break

        case LightState.YELLOW:
          await this.delay(this.YELLOW_DURATION)
          if (this.active) {
            this.currentState = LightState.RED
            this.lastStateChange = Date.now()
            this.log(`Traffic light changed to ${this.currentState}`)
          }
          break
      }

      if (this.active) {
        updateState()
      }
    }

    updateState()
  }

  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => {
      this.stateUpdateInterval = setTimeout(resolve, ms)
    })
  }

  // Add neighbor connections
  public addNeighbor(nodeId: string): void {
    this.neighbors.add(nodeId)
  }

  // Remove neighbor connections
  public removeNeighbor(nodeId: string): void {
    this.neighbors.delete(nodeId)
    this.failedNeighbors.delete(nodeId)
    this.adjustTimings()
  }

  // Update neighbor status and adjust timings
  public updateNeighborStatus(nodeId: string, failed: boolean): void {
    if (failed) {
      this.failedNeighbors.add(nodeId)
    } else {
      this.failedNeighbors.delete(nodeId)
    }
    this.adjustTimings()
  }

  // Adjust timings based on network state
  private adjustTimings(): void {
    if (this.failedNeighbors.size === 0) {
      // Reset to base durations if no failed neighbors
      this.RED_DURATION = this.BASE_RED_DURATION
      this.GREEN_DURATION = this.BASE_GREEN_DURATION
      return
    }

    const failureRatio = this.failedNeighbors.size / this.neighbors.size
    
    // Adjust green time based on number of failed neighbors
    // More failed neighbors = longer green time to handle increased load
    this.GREEN_DURATION = Math.min(
      this.BASE_GREEN_DURATION * (1 + failureRatio),
      20000 // Cap at 20 seconds
    )

    // Reduce red time proportionally to keep cycle time reasonable
    this.RED_DURATION = Math.max(
      this.BASE_RED_DURATION * (1 - failureRatio * 0.5),
      5000 // Minimum 5 seconds for safety
    )

    this.log(`Adjusted timings - Red: ${this.RED_DURATION}ms, Green: ${this.GREEN_DURATION}ms`)
  }

  public simulateFailure(): void {
    this.active = false
    this.isDisabled = true
    this.stop()
    // Notify neighbors of failure
    this.server.notifyNodeFailure(this.nodeId)
    this.log(`Node ${this.nodeId} simulated failure`)
  }

  public revive(): void {
    this.active = true
    this.isDisabled = false
    this.start()
    // Notify neighbors of revival
    this.server.notifyNodeRevival(this.nodeId)
    this.log(`Node ${this.nodeId} revived`)
  }

  public stop(): void {
    this.active = false
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval)
    }
    if (this.stateUpdateInterval) {
      clearTimeout(this.stateUpdateInterval)
    }
  }

  private log(message: string): void {
    this.logCallback(`${this.nodeId}: ${message}`)
  }
}

File: old/TrafficLightNode.java
import java.io.*;
import java.net.*;
import java.util.Scanner;

class TrafficLightNode {
    private static final String SERVER_ADDRESS = "localhost";
    private static final int SERVER_PORT = 5000;
    private static final int HEARTBEAT_INTERVAL = 3000; 
    private static final int RED_DURATION = 10000;
    private static final int GREEN_DURATION = 10000;
    private static final int YELLOW_DURATION = 3000;

    private String nodeId;
    private volatile boolean active = true;
    private Socket socket;
    private PrintWriter out;
    private LightState initialState;

    public enum LightState { RED, GREEN, YELLOW }

    public TrafficLightNode(String nodeId) {
        this.nodeId = nodeId;
        // Set initial state based on node ID to create a pattern
        if (nodeId.equals("Node1")) {
            initialState = LightState.RED;
        } else if (nodeId.equals("Node2")) {
            initialState = LightState.GREEN;
        } else if (nodeId.equals("Node3")) {
            initialState = LightState.RED;
        } else {
            // For additional nodes, set initial state based on position
            int nodeNum = Integer.parseInt(nodeId.substring(4));
            initialState = (nodeNum % 3 == 2) ? LightState.GREEN : LightState.RED;
        }
    }

    public void start() {
        try {
            socket = new Socket(SERVER_ADDRESS, SERVER_PORT);
            out = new PrintWriter(socket.getOutputStream(), true);
            System.out.println("Connected to server at " + SERVER_ADDRESS + ":" + SERVER_PORT);

            new Thread(new HeartbeatSender()).start();
            new Thread(new TrafficLightStateUpdater(initialState)).start();
            new Thread(new UserInputHandler()).start();

        } catch(IOException e) {
            e.printStackTrace();
        }
    }

    private class HeartbeatSender implements Runnable {
        public void run() {
            while (active) {
                sendHeartbeat();
                try {
                    Thread.sleep(HEARTBEAT_INTERVAL);
                } catch (InterruptedException e) {
                }
            }
            System.out.println("Heartbeat sender stopped.");
        }
    }

    private void sendHeartbeat() {
        if (out != null && active) {
            out.println("HEARTBEAT:" + nodeId);
            System.out.println("Sent heartbeat from node " + nodeId);
        }
    }

    private class TrafficLightStateUpdater implements Runnable {
        private LightState state;

        public TrafficLightStateUpdater(LightState initialState) {
            this.state = initialState;
            // Immediately report initial state
            if (out != null) {
                out.println("STATE:" + nodeId + ":" + state);
            }
            System.out.println("Node " + nodeId + ": Traffic light initialized to " + state);
        }

        public void run() {
            while (active) {
                try {
                    // First send the current state
                    if (out != null) {
                        out.println("STATE:" + nodeId + ":" + state);
                    }
                    
                    // Sleep duration depends on current state
                    switch (state) {
                        case RED:
                            Thread.sleep(RED_DURATION);
                            if (active) {
                                state = LightState.GREEN;
                            }
                            break;
                        case GREEN:
                            Thread.sleep(GREEN_DURATION);
                            if (active) {
                                state = LightState.YELLOW;
                            }
                            break;
                        case YELLOW:
                            Thread.sleep(YELLOW_DURATION);
                            if (active) {
                                state = LightState.RED;
                            }
                            break;
                    }
                    
                    // Log the state change
                    if (active) {
                        System.out.println("Node " + nodeId + ": Traffic light changed to " + state);
                    }
                } catch (InterruptedException e) {
                    if (active) {
                        Thread.currentThread().interrupt();
                        continue;
                    }
                    break;
                }
            }
            // Send final inactive state
            if (out != null) {
                out.println("INACTIVE:" + nodeId);
            }
            System.out.println("Traffic light state updater stopped.");
        }
    }

    private class UserInputHandler implements Runnable {
        public void run() {
            Scanner scanner = new Scanner(System.in);
            System.out.println("Enter 'fail' to simulate node failure for " + nodeId + ":");
            while (active) {
                String input = scanner.nextLine().trim();
                if ("fail".equalsIgnoreCase(input)) {
                    System.out.println("Simulating failure for node " + nodeId);
                    active = false;
                    try {
                        if (socket != null) {
                            socket.close();
                        }
                    } catch(IOException e) {
                    }
                }
            }
            scanner.close();
        }
    }

    public static void main(String[] args) {
        String nodeId = args.length > 0 ? args[0] : "Node1";
        TrafficLightNode node = new TrafficLightNode(nodeId);
        node.start();
    }
}
